crates/services/upgradable-executor/wasm-executor/src/ext.rs

```diff
@@ -0,0 +1,281 @@
+pub fn next_transactions(gas_limit: u64) -> anyhow::Result<Vec<MaybeCheckedTransaction>> {
+    let next_size = unsafe { host::peek_next_txs_size(gas_limit) };
+
+    if next_size == 0 {
+        return Ok(Vec::new());
+    }
+
+    let mut encoded_block = vec![0u8; next_size as usize];
+    let size = encoded_block.len();
+
+    unsafe {
+        host::consume_next_txs(
+            Ptr32Mut::from_slice(encoded_block.as_mut_slice()),
+            u32::try_from(size).expect("We only support wasm32 target; qed"),
+        )
+    };
+
+    let txs: Vec<Transaction> = postcard::from_bytes(&encoded_block)
+        .map_err(|e| anyhow::anyhow!("Failed to decode the transactions: {:?}", e))?;
+
+    Ok(txs
+        .into_iter()
+        .map(MaybeCheckedTransaction::Transaction)
+        .collect())
+}
+
+/// Gets the size of the value under the `key` in the `column` from the storage.
+pub fn size_of_value(key: &[u8], column: u32) -> anyhow::Result<Option<usize>> {
+    let val = unsafe {
+        host::storage_size_of_value(
+            Ptr32::from_slice(key),
+            u32::try_from(key.len()).expect("We only support wasm32 target; qed"),
+            column,
+        )
+    };
+
+    let (exists, size, result) = unpack_exists_size_result(val);
+
+    if result != 0 {
+        return Err(anyhow::anyhow!(
+            "Failed to get the size of the value from the WASM storage"
+        ));
+    }
+
+    if exists {
+        Ok(Some(size as usize))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Gets the value under the `key` in the `column` from the storage and copies it to the `out`.
+pub fn get(key: &[u8], column: u32, out: &mut [u8]) -> anyhow::Result<()> {
+    let output_size = out.len();
+    let result = unsafe {
+        host::storage_get(
+            Ptr32::from_slice(key),
+            u32::try_from(key.len()).expect("We only support wasm32 target; qed"),
+            column,
+            Ptr32Mut::from_slice(out),
+            u32::try_from(output_size).expect("We only support wasm32 target; qed"),
+        )
+    };
+
+    if result.0 != 0 {
+        return Err(anyhow::anyhow!(
+            "Failed to get the value from the WASM storage"
+        ));
+    }
+
+    Ok(())
+}
+
+/// Returns `true` if the relayer is enabled. Uses a host function.
+pub fn relayer_enabled() -> bool {
+    unsafe { host::relayer_enabled() }
+}
+
+/// Gets the events at the `da_block_height` from the relayer by using host functions.
+pub fn relayer_get_events(da_block_height: DaBlockHeight) -> anyhow::Result<Vec<Event>> {
+    let size = unsafe { host::relayer_size_of_events(da_block_height.into()) };
+    let (exists, size, result) = unpack_exists_size_result(size);
+
+    if result != 0 {
+        return Err(anyhow::anyhow!(
+            "Failed to get the size of events from the WASM relayer"
+        ));
+    }
+
+    if !exists || size == 0 {
+        return Ok(vec![]);
+    }
+
+    let mut encoded_events = vec![0u8; size as usize];
+
+    unsafe {
```

Comment:
Generally it's good practice to add some comments to unsafe code to explain why it's marked unsafe and any precautions taken to ensure no memory corruption occcurs or similar justifications

Reply:
The function is unsafe because we call a method from `extern`. I'm not sure what comment should say in that case=) "We assume that the host implements corresponding function overwise it should fail before we even started execution of the WASM."